File: src/main/java/com/DavidM1A2/AfraidOfTheDark/client/gui/guiScreens/SextantGUI.java
/*
 * Author: David Slovikosky 
 * Mod: Afraid of the Dark 
 * Ideas and Textures: Michael Albertson
 */
package com.DavidM1A2.AfraidOfTheDark.client.gui.guiScreens;

import java.awt.Color;
import java.io.IOException;
import java.util.Random;

import org.lwjgl.opengl.GL11;

import com.DavidM1A2.AfraidOfTheDark.AfraidOfTheDark;
import com.DavidM1A2.AfraidOfTheDark.client.gui.AOTDActionListener;
import com.DavidM1A2.AfraidOfTheDark.client.gui.baseControls.AOTDGuiButton;
import com.DavidM1A2.AfraidOfTheDark.client.gui.baseControls.AOTDGuiComponent;
import com.DavidM1A2.AfraidOfTheDark.client.gui.baseControls.AOTDGuiImage;
import com.DavidM1A2.AfraidOfTheDark.client.gui.baseControls.AOTDGuiPanel;
import com.DavidM1A2.AfraidOfTheDark.client.gui.baseControls.AOTDGuiScreen;
import com.DavidM1A2.AfraidOfTheDark.client.gui.baseControls.AOTDGuiTextField;
import com.DavidM1A2.AfraidOfTheDark.client.gui.baseControls.TextAlignment;
import com.DavidM1A2.AfraidOfTheDark.client.settings.ClientData;
import com.DavidM1A2.AfraidOfTheDark.common.packets.SpawnMeteor;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.BlockPos;
import net.minecraft.util.ChatComponentText;

public class SextantGUI extends AOTDGuiScreen
{
	private AOTDGuiTextField angle;
	private AOTDGuiTextField latitude;
	private AOTDGuiTextField longitude;

	public SextantGUI()
	{
		AOTDGuiPanel background = new AOTDGuiPanel((640 - 256) / 2, (360 - 256) / 2, 256, 256, false);

		AOTDGuiImage backgroundImage = new AOTDGuiImage(0, 0, 256, 256, "textures/gui/sextant.png");
		background.add(backgroundImage);

		this.angle = new AOTDGuiTextField(15, 108, 120, 30, ClientData.getTargaMSHandFontSized(45f));
		this.latitude = new AOTDGuiTextField(15, 140, 120, 30, ClientData.getTargaMSHandFontSized(45f));
		this.longitude = new AOTDGuiTextField(15, 172, 120, 30, ClientData.getTargaMSHandFontSized(45f));
		angle.setTextColor(Color.WHITE);
		angle.setGhostText("Angle");
		background.add(angle);
		latitude.setTextColor(Color.WHITE);
		latitude.setGhostText("Latitude");
		background.add(latitude);
		longitude.setTextColor(Color.WHITE);
		longitude.setGhostText("Longitude");
		background.add(longitude);

		AOTDGuiButton confirm = new AOTDGuiButton(15, 204, 120, 20, ClientData.getTargaMSHandFontSized(40f), "afraidofthedark:textures/gui/signButton.png", "afraidofthedark:textures/gui/signButtonHovered.png");
		confirm.setText("Calculate");
		confirm.setTextAlignment(TextAlignment.ALIGN_CENTER);
		confirm.addActionListener(new AOTDActionListener()
		{
			@Override
			public void actionPerformed(AOTDGuiComponent component, ActionType actionType)
			{
				if (actionType == actionType.MousePressed)
					if (component.isHovered())
						try
						{
							if ((Integer.parseInt(SextantGUI.this.angle.getText()) == ClientData.selectedMeteor[0]) && (Integer.parseInt(SextantGUI.this.latitude.getText()) == ClientData.selectedMeteor[1]) && (Integer.parseInt(SextantGUI.this.longitude.getText()) == ClientData.selectedMeteor[2]))
							{
								SextantGUI.this.tellServerToCreateMeteor(entityPlayer);
								entityPlayer.closeScreen();
							}
							else if (SextantGUI.this.angle.getText().isEmpty() || SextantGUI.this.latitude.getText().isEmpty() || SextantGUI.this.longitude.getText().isEmpty())
								entityPlayer.addChatMessage(new ChatComponentText("I forgot to fill out one of the fields."));
							else
								entityPlayer.addChatMessage(new ChatComponentText("The calculation was not sucessful.\nMaybe I entered incorrect numbers or should find another meteor to track."));
						}
						catch (final Exception e)
						{
							entityPlayer.addChatMessage(new ChatComponentText("The calculation was not sucessful.\nMaybe I entered incorrect numbers or should find another meteor to track."));
						}
			}
		});
		background.add(confirm);

		this.getContentPane().add(background);
	}

	// If E is typed we close the GUI screen
	@Override
	protected void keyTyped(final char character, final int keyCode) throws IOException
	{
		super.keyTyped(character, keyCode);
		if (!this.angle.isFocused() && !this.latitude.isFocused() && !this.longitude.isFocused())
		{
			if (keyCode == INVENTORY_KEYCODE)
			{
				entityPlayer.closeScreen();
				GL11.glFlush();
			}
		}
	}

	private void tellServerToCreateMeteor(final EntityPlayer entityPlayer)
	{
		final Random random = entityPlayer.worldObj.rand;
		final int xLocOfDrop = (int) entityPlayer.posX + (((random.nextDouble() >= .5) ? -1 : 1) * (entityPlayer.worldObj.rand.nextInt(500) + 15));
		final int zLocOfDrop = (int) entityPlayer.posZ + (((random.nextDouble() >= .5) ? -1 : 1) * (entityPlayer.worldObj.rand.nextInt(500) + 15));

		final BlockPos location = new BlockPos(xLocOfDrop, 255, zLocOfDrop);
		AfraidOfTheDark.getPacketHandler().sendToServer(new SpawnMeteor(location, 3, 3, ClientData.watchedMeteorType.getIndex()));
		entityPlayer.addChatMessage(new ChatComponentText("Based off of this information the meteor fell at " + xLocOfDrop + ", " + zLocOfDrop));
		ClientData.selectedMeteor = new int[]
		{ -1, -1, -1 };
		ClientData.watchedMeteorType = null;
	}

	@Override
	public boolean inventoryToCloseGuiScreen()
	{
		return false;
	}

	@Override
	public boolean drawGradientBackground()
	{
		return true;
	}
}

File: src/main/java/com/davidm1a2/afraidofthedark/client/gui/guiScreens/SextantGUI.kt
package com.davidm1a2.afraidofthedark.client.gui.guiScreens

import com.davidm1a2.afraidofthedark.AfraidOfTheDark
import com.davidm1a2.afraidofthedark.client.gui.base.AOTDGuiScreen
import com.davidm1a2.afraidofthedark.client.gui.base.TextAlignment
import com.davidm1a2.afraidofthedark.client.gui.events.AOTDMouseEvent
import com.davidm1a2.afraidofthedark.client.gui.standardControls.AOTDGuiButton
import com.davidm1a2.afraidofthedark.client.gui.standardControls.AOTDGuiImage
import com.davidm1a2.afraidofthedark.client.gui.standardControls.AOTDGuiPanel
import com.davidm1a2.afraidofthedark.client.gui.standardControls.AOTDGuiTextField
import com.davidm1a2.afraidofthedark.client.settings.ClientData
import com.davidm1a2.afraidofthedark.common.constants.Constants
import com.davidm1a2.afraidofthedark.common.constants.LocalizationConstants
import com.davidm1a2.afraidofthedark.common.packets.otherPackets.ProcessSextantInput
import net.minecraft.util.text.TextComponentTranslation
import org.lwjgl.util.Color

/**
 * Gui screen that represents the sextant GUI
 *
 * @constructor initializes the GUI
 * @property angle The text field containing the meteor's drop angle
 * @property latitude The text field containing the meteor's latitude
 * @property longitude The text field containing the meteor's longitude
 */
class SextantGUI : AOTDGuiScreen() {
    private val angle: AOTDGuiTextField
    private val latitude: AOTDGuiTextField
    private val longitude: AOTDGuiTextField

    init {
        // The gui will be 256x256
        val guiSize = 256

        // Background panel holds all the gui items
        val background =
            AOTDGuiPanel((Constants.GUI_WIDTH - guiSize) / 2, (Constants.GUI_HEIGHT - guiSize) / 2, 256, 256, false)

        // Add an image to the background of the sextant texture
        val backgroundImage = AOTDGuiImage(0, 0, guiSize, guiSize, "afraidofthedark:textures/gui/telescope/sextant.png")
        background.add(backgroundImage)

        // Grab the font for the text fields
        val textFieldFont = ClientData.getOrCreate(45f)

        // Initialize fields
        angle = AOTDGuiTextField(15, 108, 120, 30, textFieldFont)
        latitude = AOTDGuiTextField(15, 140, 120, 30, textFieldFont)
        longitude = AOTDGuiTextField(15, 172, 120, 30, textFieldFont)

        // All fields are white and contain ghost text based on what they represent
        angle.setTextColor(Color(255, 255, 255))
        angle.setGhostText("Angle")
        background.add(angle)
        latitude.setTextColor(Color(255, 255, 255))
        latitude.setGhostText("Latitude")
        background.add(latitude)
        longitude.setTextColor(Color(255, 255, 255))
        longitude.setGhostText("Longitude")
        background.add(longitude)

        // Create a calculate button that performs the math and returns drop location coordinates
        val confirm = AOTDGuiButton(
            15,
            204,
            120,
            20,
            "afraidofthedark:textures/gui/journal_sign/sign_button.png",
            "afraidofthedark:textures/gui/journal_sign/sign_button_hovered.png",
            ClientData.getOrCreate(40f)
        )
        // Text just says calculate
        confirm.setText("Calculate")
        // Center the text
        confirm.setTextAlignment(TextAlignment.ALIGN_CENTER)
        // When clicked tell the server to validate the numbers and create a meteor if possible
        confirm.addMouseListener {
            if (it.eventType == AOTDMouseEvent.EventType.Click) {
                // Ensure this button was the one clicked
                if (it.source.isHovered && it.clickedButton == AOTDMouseEvent.LEFT_MOUSE_BUTTON) {
                    // Grab the text fron the text fields
                    val dropAngleText = angle.getText()
                    val latitudeText = latitude.getText()
                    val longitudeText = longitude.getText()
                    // If any of the fields are empty print a message
                    if (dropAngleText.isEmpty() || latitudeText.isEmpty() || longitudeText.isEmpty()) {
                        entityPlayer.sendMessage(TextComponentTranslation(LocalizationConstants.Sextant.FIELD_EMPTY))
                    }

                    // If any field is invalid send the player an error, otherwise send the info to the server
                    try {
                        AfraidOfTheDark.INSTANCE.packetHandler.sendToServer(
                            ProcessSextantInput(
                                dropAngleText.toInt(),
                                latitudeText.toInt(),
                                longitudeText.toInt()
                            )
                        )
                    } catch (e: NumberFormatException) {
                        entityPlayer.sendMessage(TextComponentTranslation(LocalizationConstants.Sextant.INVALID_VALUE))
                    }
                    entityPlayer.closeScreen()
                }
            }
        }
        background.add(confirm)
        contentPane.add(background)
    }

    /**
     * @return True if none of the 3 fields are focused, false otherwise
     */
    override fun inventoryToCloseGuiScreen(): Boolean {
        return !angle.isFocused && !latitude.isFocused && !longitude.isFocused
    }

    /**
     * @return True since this UI uses a gradient background
     */
    override fun drawGradientBackground(): Boolean {
        return true
    }
}